api/accounts/changeUsername.ts

```diff
@@ -0,0 +1,24 @@
+import handler from '../handler';
+import withConnection from '../db';
+import * as auth from '../authenticate';
+import * as errors from '../errors';
+import { Client } from 'pg';
+
+// Expects 'username'
+// Returns the randomly generated password
+export default handler(async (req, userId): Promise<string> => {
+    errors.requireParams(req.body, ['username','newusername'],);
+    const username = req.body.username;
+    const newname = req.body.newusername;
+    return withConnection(async (db: Client) => {
+        const matchingAccount = await db.query('SELECT username FROM authority WHERE username = $1', [username]);
+        if (matchingAccount.rows.length !== 1) {
+            throw new errors.BadParameter('No such account with username ' + username);
+        }
+        const updateResults = await db.query('UPDATE authority SET username = $1 WHERE username = $2', [newname, username]);
+        if (updateResults.rowCount !== 1){
+            throw new errors.BadParameter('No such account with username ' + username); 
```

Comment:
This error is misleading. Instead throw an APIError that indicates that this should never happen ("Somehow failed to...")

Reply:
Not quite sure what you mean by APIError?

Reply:
(I clarified with Bar: I think we should remove the `matchingAccount` SQL call and check and just use this check to throw the bad parameter error.